CommunityToolkit.Mvvm/Input/RelayCommand{T}.cs

```diff
@@ -100,6 +107,66 @@ public void Execute(T? parameter)
     /// <inheritdoc/>
     public void Execute(object? parameter)
     {
-        Execute((T?)parameter);
+        if (!TryGetCommandArgument(parameter, out T? result))
+        {
+            ThrowArgumentExceptionForInvalidCommandArgument(parameter);
+        }
+
+        Execute(result);
+    }
+
+    /// <summary>
+    /// Tries to get a command argument of compatible type <typeparamref name="T"/> from an input <see cref="object"/>.
+    /// </summary>
+    /// <param name="parameter">The input parameter.</param>
+    /// <param name="result">The resulting <typeparamref name="T"/> value, if any.</param>
+    /// <returns>Whether or not a compatible command argument could be retrieved.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static bool TryGetCommandArgument(object? parameter, out T? result)
+    {
+        // If the argument is null and the default value of T is also null, then the
+        // argument is valid. T might be a reference type or a nullable value type.
+        if (parameter is null && default(T) is null)
+        {
+            result = default;
+
+            return true;
+        }
+
+        // Check if the argument is a T value, so either an instance of a type or a derived
+        // type of T is a reference type, an interface implementation if T is an interface,
+        // or a boxed value type in case T was a value type.
+        if (parameter is T argument)
+        {
+            result = argument;
+
+            return true;
+        }
+
+        result = default;
+
+        return false;
+    }
+
+    /// <summary>
+    /// Throws an <see cref="ArgumentException"/> if an invalid command argument is used.
+    /// </summary>
+    /// <param name="parameter">The input parameter.</param>
+    /// <exception cref="ArgumentException">Thrown with an error message to give info on the invalid parameter.</exception>
+    [DoesNotReturn]
+    internal static void ThrowArgumentExceptionForInvalidCommandArgument(object? parameter)
+    {
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Is `NoInlining` intended here?

Reply:
Yeah it's to make sure the JIT doesn't mess up handling the caller as a throw helper.